### ISSUE TITLE
Better build system?

### ISSUE THREAD

**Issue by LaserEyess** [snesrev/sm#13]
It looks like you're using two separate, hand-rolled build systems here. I understand this project is in its infancy but I think you'll want to use a "real" build system sooner rather than later. The two major options for cross-platform C/C++ projects are[ CMake](https://cmake.org/cmake/help/latest/) and [meson](https://mesonbuild.com/). Both have their pros and cons, and I have experience with both. 

In general, I would recommend meson, but visual studio support (if that's a hard requirement?) is less mature than CMake. I'd be willing to make a PR for either, as I'd like to start hacking on this project but I also don't want to deal with makefiles.

**Comment by shadow42085** [comment]
frankly I would ditch Visual Studio support for msys2/ttc since you can use any lightweight C/C++ IDE with it and visual studio is very bloated where as Msys2 is not as bloated since it's just the compiler

**Comment by eli-schwartz** [comment]
Visual Studio is an IDE, but MSVC is the compiler that comes bundled inside it. Using the latter is probably more compelling than the former.



> In general, I would recommend meson, but visual studio support (if that's a hard requirement?) is less mature than CMake.

The Visual Studio + MSVC support in Meson may not be as high quality as the ninja + MSVC support, but those sorts of issues are most likely only going to show up in projects that venture into things a lot more complicated than just C/C++. Anyway, the solution is of course to use ninja anyway -- it will always be much, much faster. :D

**Comment by LaserEyess** [comment]
I agree, but let me clarify that by "visual studio" I mean visual studio project files built with msbuild.exe, not the IDE. I figure mingw+ninja should be enough on Windows but sometimes Windows development leads to... strange dependencies.

**Comment by jason-oliveira** [comment]
The existing build system works great with a linux-based workflow (MSVC is gross), but CMake is always a plus.

**Comment by snesrev** [comment]
We will not ditch VS because VS is what I use for development.

**Comment by LaserEyess** [comment]
@snesrev is this workflow acceptable to you? https://learn.microsoft.com/en-us/cpp/build/cmake-projects-in-visual-studio?view=msvc-170

If so, I can make a PR that ports the build system to cmake.

**Comment by snesrev** [comment]
@LaserEyess yes it looks ok
